### PR TITLE
drivers: display: ili9340 refactoring

### PR DESCRIPTION
**Pull request by pavlohamov**
- extend ili9340 driver to use config from dts
 - add posbility to use single driver for ili9xxx displays
 - create generic sheild for the ili934x based boards

### CODE REVIEW

File: boards/shields/ili9340_generic/ili9340_adafruit_2_8_tftv2.overlay [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+		reg = <0>;
+		cmd-data-gpios = <&arduino_header 15 GPIO_ACTIVE_LOW>;	/* D9 */
+		type = "ili9340";
+		height = <320>;
+		width = <240>;
+		x-offset = <0>;
+		y-offset = <0>;
+		pixel-format = "RGB888";
+
+	    power-ctrl-1 = [23];
```
Comment (vanwinkeljan) [inline boards/shields/ili9340_generic/ili9340_adafruit_2_8_tftv2.overlay]: Keep indentation the same as above (2 tabs)

File: boards/shields/ili9340_generic/ili9340_seeed_2_8_tftv2.overlay [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+		reg = <0>;
+		cmd-data-gpios = <&arduino_header 15 GPIO_ACTIVE_LOW>;	/* D9 */
+		type = "ili9341";
+		height = <320>;
+		width = <240>;
+		x-offset = <0>;
+		y-offset = <0>;
+		pixel-format = "RGB888";
+
+	    power-ctrl-a = [39 2C 00 34 02];
```
Comment (vanwinkeljan) [inline boards/shields/ili9340_generic/ili9340_seeed_2_8_tftv2.overlay]: Keep indentation the same as above (2 tabs)

File: drivers/display/display_ili9340.c [hunk 3]
```diff
@@ -1,154 +1,152 @@
...
-
-#define ILI9340_CMD_DATA_PIN_COMMAND 1
-#define ILI9340_CMD_DATA_PIN_DATA 0
+#include <logging/log.h>
+LOG_MODULE_REGISTER(display_ili9340, CONFIG_DISPLAY_LOG_LEVEL);
 
-/* The number of bytes taken by a RGB pixel */
-#ifdef CONFIG_ILI9340_RGB565
-#define ILI9340_RGB_SIZE 2U
+#if (DT_INST_0_ILITEK_ILI9XXX_TYPE_ENUM == 0)
```
Comment (vanwinkeljan) [inline drivers/display/display_ili9340.c]: Please remove this code `DT_INST_0_ILITEK_ILI9XXX_TYPE_ENUM` can never be 0 as it is required device parameter.

File: drivers/display/display_ili9340.c [hunk 4]
```diff
@@ -1,154 +1,152 @@
...
-	LOG_DBG("Initializing display driver");
+#define DEV_TYPE DT_INST_0_ILITEK_ILI9XXX_TYPE
 
-	data->spi_dev = device_get_binding(DT_INST_0_ILITEK_ILI9340_BUS_NAME);
-	if (data->spi_dev == NULL) {
-		LOG_ERR("Could not get SPI device for ILI9340");
-		return -EPERM;
-	}
+#if (RES_VER > ILI9XXX_Y_MAX) && (RES_VER > ILI9XXX_X_MAX)
+#error "Please check height of \"ilitek,"DEV_TYPE""\" device"
```
Comment (vanwinkeljan) [inline drivers/display/display_ili9340.c]: Better be consistent in error messages; `"ilitek,"DEV_TYPE"` VS `ilitek,ili9340` bellow

**Review by vanwinkeljan — CHANGES_REQUESTED** [review]
In general this PR looks good to me except some minor remarks.

Could you also update the display sample yaml file so that CI passes?

File: drivers/display/display_ili9340.c [hunk 5]
```diff
@@ -185,120 +183,307 @@ static int ili9340_read(const struct device *dev, const u16_t x,
...
-	capabilities->current_pixel_format = PIXEL_FORMAT_RGB_888;
+	capabilities->x_resolution = RES_HOR;
+	capabilities->y_resolution = RES_VER;
+	capabilities->supported_pixel_formats =
+			PIXEL_FORMAT_RGB_565 | PIXEL_FORMAT_RGB_888;
+	capabilities->current_pixel_format = data->pixel_format;
+	capabilities->current_orientation = data->orientation;
+}
+
+#define EXEC_CMD(cmd) { \
```
Comment (jfischer-phytec-iot) [inline drivers/display/display_ili9340.c]: This would be hard to debug, return is opaque to the reader. `_` must not be used and anyway name should be ili93xx_exec_cmd(cmd).

File: drivers/display/display_ili9340.h [hunk 6]
```diff
@@ -9,67 +10,45 @@
...
-void ili9340_transmit(struct ili9340_data *data, u8_t cmd, void *tx_data,
-		      size_t tx_len);
-
-/**
- * Perform LCD specific initialization
- *
- * @param data Device data structure
- */
-void ili9340_lcd_init(struct ili9340_data *data);
+#define ILI9XXX_CMD_SOFTWARE_RESET 0x01
```
Comment (jfischer-phytec-iot) [inline drivers/display/display_ili9340.h]: Align constants with tabs

File: dts/bindings/display/ilitek,ili9xxx.yaml [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        The D/CX pin of ILI9340 is active low (transmission command byte).
+        If connected directly the MCU pin should be configured
+        as active low.
+
+    pixel-format:
+      type: string
+      required: false
+      description: pixel data format
+      enum:
+        - "RGB565"
```
Comment (jfischer-phytec-iot) [inline dts/bindings/display/ilitek,ili9xxx.yaml]: There is still Kconfig menu for that, either DT or Kconfig

**Review by jfischer-no — CHANGES_REQUESTED** [review]
This patch has to be spited in self contained logical commits, like:
```
- drivers: display_ili9340: fix style, cleanup
- drivers: display_ili9340: fix style, move init function to the bottom
- drivers: display_ili9340: rename to display_ili93xx
- shields: rename/move adafruit_2_8_tft_touch_v2 shield to ili93xx_generic
- drivers: display_ili93xx: obtain resolution and offsets from DT
- drivers: display_ili93xx: obtain panel settings and parameters from DT
- drivers: display_ili93xx: add support for ABCD controller
- shields: ili93xx_generic: add support for ABCD display
```

File: drivers/display/display_ili9340.c (hunk 3 above)
Comment (pavlohamov) [reply drivers/display/display_ili9340.c]: As far as i understood how enums in dts works they are numerated in the same way as in C. 
0 - stand for the firts item in the dts enum (which corresponds to ili9163 for this case).
But this is too implicit. string comparison wont work for #if statement. I will try to investiagate what could be done with this magic number

**Comment by pavlohamov** [comment]
@vanwinkeljan @jfischer-phytec-iot  Please excuse some of obvious thins like typo and indent. I wanted to get some feedback about solution in general. Thank you for your comments.

**Comment by jfischer-phytec-iot** [comment]
> @vanwinkeljan @jfischer-phytec-iot Please excuse some of obvious thins like typo and indent.

No worry that is what review is for.

File: drivers/display/display_ili9340.c (hunk 3 above)
Comment (vanwinkeljan) [reply drivers/display/display_ili9340.c]: OK I did not know that the DTS enums returned numbers instead of strings, forget the remark

File: drivers/display/display_ili9340.c [hunk 8]
```diff
@@ -1,154 +1,152 @@
...
-	k_sleep(K_MSEC(1));
-	gpio_pin_set(data->reset_gpio,
-		     DT_INST_0_ILITEK_ILI9340_RESET_GPIOS_PIN, 1);
-	k_sleep(K_MSEC(1));
-	gpio_pin_set(data->reset_gpio,
-		     DT_INST_0_ILITEK_ILI9340_RESET_GPIOS_PIN, 0);
-	k_sleep(K_MSEC(5));
-#endif
+	gpio_pin_set(data->cmd,
+			DT_INST_0_ILITEK_ILI9XXX_CMD_DATA_GPIOS_PIN, 0);
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9340.c]: Can't agree with that. From datasheet:
> The D/CX is the signal of data/command
> selection. When D/CX=’1’, D [17:0] bits are display RAM data or command’s parameters. When D/CX=’0’, D
> [17:0] bits are commands.
But i understand your confusion. variable is called "cmd". I will rename int to data or ncmd (now shure which option is better)

File: drivers/display/display_ili9340.c [hunk 9]
```diff
@@ -1,154 +1,152 @@
...
+		return rv;
+	}
 
-	LOG_DBG("Initializing LCD");
-	ili9340_lcd_init(data);
+	if (tx_data != NULL) {
+		tx_buf.buf = tx_data;
+		tx_buf.len = tx_len;
+		gpio_pin_set(data->cmd,
+			       DT_INST_0_ILITEK_ILI9XXX_CMD_DATA_GPIOS_PIN, 1);
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9340.c]: > The D/CX is the signal of data/command
> selection. When D/CX=’1’, D [17:0] bits are display RAM data or command’s parameters. When D/CX=’0’, D
> [17:0] bits are commands.
 But variable definitely has to be renamed

File: drivers/display/display_ili9340.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply drivers/display/display_ili9340.c]: No,  `DT_INST_0_ILITEK_ILI9XXX_CMD_DATA_GPIOS_PIN, 1);` is correct because [The D/CX pin of ILI9340 is active low (transmission command byte).](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/display/ilitek%2Cili9340.yaml). No need for ncmd or cmd.

File: drivers/display/display_ili9340.c [hunk 10]
```diff
@@ -185,120 +183,307 @@ static int ili9340_read(const struct device *dev, const u16_t x,
...
+		LOG_ERR("CMD "#cmd" failed %d", _rv); \
+		return _rv; \
+	} \
+}
+
+static int ili9xxx_lcd_init(struct device *dev)
+{
+	struct ili9xxx_data *data = (struct ili9xxx_data *)dev->driver_data;
+
+#ifdef DT_INST_0_ILITEK_ILI9XXX_POWER_CTRL_A
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/display/display_ili9340.c]: Done

File: drivers/display/display_ili9340.c (hunk 5 above)
Comment (pavlohamov) [reply drivers/display/display_ili9340.c]: Done

**Comment by pavlohamov** [comment]
i've spitted single commit to multiple, Fixing comment during this process. Also i've removed addition of new controller type. I want to do it in separate PR.
@vanwinkeljan @jfischer-phytec-iot Could you be so kind to take a look?

**Review by vanwinkeljan — APPROVED** [review]

**Comment by pavlohamov** [comment]
@jfischer-phytec-iot could you be so kind to take a look?

**Comment by pavlohamov** [comment]
Rebased to latest master

**Review by jfischer-no — CHANGES_REQUESTED** [review]
> @jfischer-phytec-iot could you be so kind to take a look?

Sorry for delay, I need few more days for review/test.

File: drivers/display/display_ili9340.c [hunk 11]
```diff
@@ -8,37 +8,31 @@
...
 
+#define RES_X_MAX 320U
+#define RES_Y_MAX 240U
+
 struct ili9340_data {
 #ifdef DT_INST_0_ILITEK_ILI9340_RESET_GPIOS_CONTROLLER
-	struct device *reset_gpio;
+	struct device *rst;
 #endif
-	struct device *command_data_gpio;
```
Comment (jfischer-phytec-iot) [inline drivers/display/display_ili9340.c]: I agree that it was to long and you probably changed it to resolve coding style issues. But it would be better to keep reset_gpio + cs_gpio, and change command_data_gpio to cmd_gpio.

You put to many different changes in one commit. Sorry, but it is hard to review.

File: drivers/display/display_ili9340.c [hunk 12]
```diff
@@ -62,46 +65,46 @@ static int ili9340_init(struct device *dev)
...
-	if (data->command_data_gpio == NULL) {
+	if (data->cmd == NULL) {
 		LOG_ERR("Could not get GPIO port for ILI9340 command/data");
 		return -EPERM;
 	}
 
-	gpio_pin_configure(data->command_data_gpio, DT_INST_0_ILITEK_ILI9340_CMD_DATA_GPIOS_PIN,
-			   GPIO_OUTPUT |
+	gpio_pin_configure(data->cmd, DT_INST_0_ILITEK_ILI9340_CMD_DATA_GPIOS_PIN,
+			   GPIO_OUTPUT_ACTIVE | GPIO_ACTIVE_LOW |
```
Comment (jfischer-phytec-iot) [inline drivers/display/display_ili9340.c]: Not necessary to change it, the flag like GPIO_ACTIVE_LOW would be obtained from *_GPIOS_FLAGS

File: drivers/display/display_ili9340.c [hunk 13]
```diff
@@ -8,37 +8,31 @@
...
 #include <logging/log.h>
-LOG_MODULE_REGISTER(display_ili9340);
+LOG_MODULE_REGISTER(display_ili9340, CONFIG_DISPLAY_LOG_LEVEL);
 
 #include <drivers/gpio.h>
 #include <sys/byteorder.h>
 #include <drivers/spi.h>
 #include <string.h>
 
+#define RES_X_MAX 320U
```
Comment (jfischer-phytec-iot) [inline drivers/display/display_ili9340.c]: This change should be in own commit and use DT height and width properties.

File: drivers/display/display_ili9xxx.c [hunk 14]
```diff
@@ -16,18 +16,50 @@ LOG_MODULE_REGISTER(display_ili9xxx, CONFIG_DISPLAY_LOG_LEVEL);
...
 #include <string.h>
 
-#define RES_X_MAX 320U
-#define RES_Y_MAX 240U
+#define ILI9XXX_X_MAX 320U
+#define ILI9XXX_Y_MAX 240U
+#define ILI9XXX_PIX_FMT_MASK \
+	(PIXEL_FORMAT_RGB_888 | PIXEL_FORMAT_RGB_565 | PIXEL_FORMAT_BGR_565)
 
+#define DT_CS_PORT        DT_INST_0_ILITEK_ILI9XXX_CS_GPIOS_CONTROLLER
```
Comment (jfischer-phytec-iot) [inline drivers/display/display_ili9xxx.c]: It does not belong in this commit, please try to move all of the changes regarding DT and style to one commit or wait for #23245

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/display/display_ili9340.c (hunk 13 above)
Comment (pavlohamov) [reply drivers/display/display_ili9340.c]: This is not related to DT config. This high bound for the ili9xxx drivers.

File: drivers/display/display_ili9340.c (hunk 12 above)
Comment (pavlohamov) [reply drivers/display/display_ili9340.c]: Make sense. My idea was to "force' inversion, but with this change nobody will able to change. thnx, will remove

File: drivers/display/display_ili9340.c (hunk 11 above)
Comment (pavlohamov) [reply drivers/display/display_ili9340.c]: Form my point of view it is excessive, vars are used with gpio_* functions indicating what kind of device is it. In general in makes sense, but here we have a particular driver where "cmd" is name of pin (cnd - command not data whoul be even better) .

So sorry, there are a lot of changes, which depend on each other

File: drivers/display/display_ili9xxx.c (hunk 14 above)
Comment (pavlohamov) [reply drivers/display/display_ili9xxx.c]: Could you be so kind to explain why? This is done inside of "drivers: display_ili9xxx: obtain display settings from DT"

**Comment by gmarull** [comment]
@pavlohamov any interest in moving this PR further? I'd be happy to take on this work as I'd like to improve ILI9340 driver and it seems that you already have done multiple of the necessary improvements.

**Comment by pavlohamov** [comment]
@gmarull, thank you for taking interest into this PR. Unfortunate I have no time for this, please take all my changes and modify them as you will

**Comment by jfischer-phytec-iot** [comment]
Close as it is superfluous after #28410 has been merged.
